src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1268,6 +1274,21 @@ public void ReportApplicationError(Exception ex)
 
         public void Advance(int bytes)
         {
+            if (bytes < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bytes));
+            }
+            else if (bytes > 0)
+            {
+                _hasAdvanced = true;
+            }
+
+            if (_requireGetMemoryNextCall)
```

Comment:
What would happen if I tried to call Advance without ever calling GetMemory?

Reply:
No matter what that is invalid, but with this new logic, Advance wouldn't be called on the PipeWriter. Agree that it is a bit confusing. Thoughts on how to improve it?

Reply:
But what error do you get? If it gets to here this won't trigger because the default value is false. You could change the default.

Reply:
Yeah I'll make it true by default